### PR TITLE
[Automation] Add Patient Search Tests

### PR DESCRIPTION
**Pull request by drryanjames**
The goal of these changes is to automate [patient search tests](https://docs.google.com/document/d/1hihKCjCQVprju_SCi1nqXCYchY-ThX4iqHPkv4asnt0/edit#heading=h.gpe6dgjlwx02).

image

The registration, clinic, lab, pharmacy, social, and vitals web pages each include the patient search component--an angular component that defines html and js used to search for a patient. Although the html and js are the same on each page we need to write automation that tests each page separately to make sure searching for patients works on every page. So we need to run the same test on different pages.

In an effort to write easily readable tests I wanted to be able to write something like the following to search for a patient:
```javascript
RegistrationPage.search("patient1")
ClinicPage.search("patient1")
LabPage.search("patient1")
PharmacyPage.search("patient1")
SocialPage.search("patient1")
VitalsPage.search("patient1")
```

However, I didn't want to have to write a separate ```search``` function in each class for each page. So I created an architecture that uses two classes: ```Page``` and ```Component```. A ```Page``` encapsulates functions and properties that make it easy to test web pages (e.g. the registration page). A ```Component``` encapsulates functions and properties that make it easy to test a component (e.g. patient search) on a web page. When a ```Page``` is created each of its ```Component```s, if any, are extended onto the page by coping the properties and functions from the ```Component``` to the ```Page```. This allows functionality in a single ```Component```, like the ```PatientSearchComponent```, to be reused on multiple ```Page```'s, like each of the pages that are tested in this PR.

The really nice part about this is that, after a ```Component```'s functionality is copied onto a ```Page```, tests can access component functionality directly on the ```Page``` object. For example, calling ```RegistrationPage.search("patient1")``` will search for the patient on the registration page using the search function that was copied from the ```PatientSearchComponent```. This allows us to keep component logic, like the ```search``` function, in a single class and reuse it on multiple pages so we can efficiently write easy to read tests.

I think the value in this architecture will be more apparent after we add more ```Component```s. I also realize there could be a better approach, so please share your thoughts and concerns.

### CODE REVIEW

File: test/e2eTests/pages/clinicPage.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const Page = require('./page')
+
+module.exports = new Page({
+  isLoaded: {
+    element: '[ng-app="clinic"]',
+    urlPart: '/clinic',
+  },
+  components: ['patientSearch'],
```
Comment (drryanjames) [inline test/e2eTests/pages/clinicPage.js]: This is how a page defines its components. In effect, this copies functionality from the ```PatientSearchComponent``` onto the clinic page object, which allows tests to call ```clinicPage.search()```

File: test/e2eTests/tests/patientSearch_test.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	validateSearch(I, Data, socialPage)
+})
+
+Scenario('Validate searching works on the vitals page', (I, Apis, Data, DashboardPage) => {
+	const vitalsPage = DashboardPage.navigateToVitalsPage()
+	validateSearch(I, Data, vitalsPage)
+})
+
+// Runs the core search scenario on the given page
+const validateSearch = (I, Data, pageWithSearch) => {
```
Comment (drryanjames) [inline test/e2eTests/tests/patientSearch_test.js]: This is what it's all been about: a single function that runs the same set of test logic on different pages. Callers navigate to the page they are testing then call this function to run the checks, so we can reuse this function for each test. ```validateSearch``` just expects the passed in ```Page``` to have search related functions, like ```seach()``` and ```seePatientRecord()```, but it doesn't know or care that they were actually copied from the ```Page```'s ```Component```

**Comment by ynurmahomed** [comment]
@drryanjames this is great.

Can we change https://github.com/esaude/esaude-emr-poc/blob/3d42215fe4557e2b5fa09727b11dd6b07f7d824f/codecept.json#L14

to `"endpoint": "http://localhost:9000/openmrs/ws/rest/v1",` to use the proxy set-up in `Gruntfile`?

Patient search is failing for me on `registration` and `pharmacy`, it looks like the app navigates to dashboard after you finish entering a patient identifier. 

I think this is the expected behaviour not sure why its not happening on the other modules, I'll check.

**Comment by ynurmahomed** [comment]
@drryanjames I usually squash commits in to one that is why your master is different from upstream.

**Comment by ynurmahomed** [comment]
It's this button:

screenshot from 2018-05-17 08-48-51

It takes all commits from your branch and turns them into a single commit, it looks better on commit history. 

This approach works better if using the [github flow](https://guides.github.com/introduction/flow/), after merge you should pull upstream master into you local master.

https://blog.github.com/2016-04-01-squash-your-commits/

**Comment by drryanjames** [comment]
Oooo, nice. I'll do that in the future.

**Pull request merged**